File: code_samples/elixir/performance_memoization.ex
# Performance: Memoization
# AI/ML Training Sample

defmodule Memoization do
  defstruct data: ""
  
  def new(), do: %Memoization{}
  
  def process(%Memoization{} = struct, input) do
    %{struct | data: input}
  end
  
  def get_data(%Memoization{data: data}), do: data
  
  def validate(%Memoization{data: data}) do
    String.length(data) > 0
  end
end

# Example usage
instance = Memoization.new()
updated = Memoization.process(instance, "example")
IO.puts("Data: " <> Memoization.get_data(updated))
IO.puts("Valid: " <> to_string(Memoization.validate(updated)))

File: code_samples/lua/performance_memoization.lua
-- Performance: Memoization
-- AI/ML Training Sample

Memoization = {}
Memoization.__ index = Memoization

function Memoization:new()
    local obj = {
        data = ""
    }
    setmetatable(obj, self)
    return obj
end

function Memoization:process(input)
    self.data = input
end

function Memoization:getData()
    return self.data
end

function Memoization:validate()
    return #self.data > 0
end

-- Example usage
local instance = Memoization:new()
instance:process("example")
print("Data: " .. instance:getData())
print("Valid: " .. tostring(instance:validate()))
